src/data_pipeline_api/file_api.py

```diff
@@ -0,0 +1,177 @@
+from io import IOBase
+from datetime import datetime
+from pathlib import Path
+from typing import Union
+from functools import wraps
+from hashlib import sha1
+import yaml
+
+from data_pipeline_api.metadata import Metadata, MetadataKey
+from data_pipeline_api.metadata_store import MetadataStore
+from data_pipeline_api.overrides import Overrides
+
+
+class FileAPI:
+    def __init__(
+        self, config_filename: Union[Path, str],
+    ):
+        config_filename = Path(config_filename)
+
+        # Compute a run_id.
+        with open(config_filename, "rb") as file:
+            m = sha1(file.read())
+        m.update(str(datetime.now()).encode())
+        self.run_id = m.hexdigest()
+
+        try:
+            with open(config_filename) as config_file:
+                config = yaml.safe_load(config_file)
+                self.data_directory = Path(config.get("data_directory", "."))
+                self.fail_on_hash_mismatch = config.get("fail_on_hash_mismatch", True)
+                self._read_overrides = Overrides(
+                    (override.get("where", {}), override.get("use", {}))
+                    for override in config.get("read", ())
+                )
+                self._write_overrides = Overrides(
+                    (override.get("where", {}), override.get("use", {}))
+                    for override in config.get("write", ())
+                )
+                self.access_log_path = Path(
+                    config.get("access_log", "access-{run_id}.yaml").format(
+                        run_id=self.run_id
+                    )
+                )
+        except Exception as exception:
+            raise ValueError("could not parse config file") from exception
+
+        if self.data_directory.is_absolute():
+            self.normalised_data_directory = self.data_directory
+        else:
+            self.normalised_data_directory = (
+                config_filename.parent / self.data_directory
+            )
+
+        if not self.access_log_path.is_absolute():
+            self.access_log_path = config_filename.parent / self.access_log_path
+        if self.access_log_path.exists():
+            raise ValueError(f"{self.access_log_path} already exists")
+
+        with open(
+            self.normalised_data_directory / "metadata.yaml"
+        ) as metadata_store_file:
+            self._metadata_store = MetadataStore(yaml.safe_load(metadata_store_file))
+
+        self._open_timestamp = datetime.now()
```

Comment:
call datetime.now() once and use it for the sha and the open timestamp? I wonder if it could be useful to know the datetime that went into the sha for any debugging etc